### PR TITLE
Ease finding slow files

### PR DESCRIPTION
**Pull request by staabm**
To ease finding the slowest files within a project, so we get a starting point what to profile I have added debug output in a similar way its done in PHPStan.

----

Run rector across your project:
```
vendor/bin/rector -vvv --debug --no-diffs | tee rector.log
```

Analyse the generated `rector.log` file with `parse.php`...
```
php parse.php
```

.. and you get a list of files sorted by the time it took rector to refactor it:

```
Slowest files
4.90 seconds: [file] packages/Testing/PHPUnit/AbstractRectorTestCase.php
4.07 seconds: [file] packages/FamilyTree/Reflection/FamilyRelationsAnalyzer.php
2.99 seconds: [file] packages/Caching/ValueObject/CacheFilePaths.php
2.95 seconds: [file] packages/BetterPhpDocParser/Attributes/AttributeMirrorer.php
2.93 seconds: [file] packages/BetterPhpDocParser/PhpDocNodeVisitor/TemplatePhpDocNodeVisitor.php
2.68 seconds: [file] packages/FamilyTree/Reflection/FamilyRelationsAnalyzer.php
2.61 seconds: [file] packages/NodeTypeResolver/TypeAnalyzer/ArrayTypeAnalyzer.php
2.53 seconds: [file] packages/PHPStanStaticTypeMapper/TypeMapper/OversizedArrayTypeMapper.php
2.07 seconds: [file] packages/BetterPhpDocParser/ValueObject/PhpDocAttributeKey.php
1.71 seconds: [file] bin/clean-phpstan.php
1.52 seconds: [file] config/set/php80.php
1.18 seconds: [file] packages/PhpAttribute/NodeAnalyzer/ExprParameterReflectionTypeCorrector.php
1.01 seconds: [file] packages/Testing/PHPUnit/AbstractRectorTestCase.php
0.97 seconds: [file] config/set/php52.php
0.89 seconds: [file] packages/BetterPhpDocParser/PhpDocNodeVisitor/TemplatePhpDocNodeVisitor.php
0.83 seconds: [file] packages/PHPStanStaticTypeMapper/TypeMapper/ResourceTypeMapper.php
0.83 seconds: [file] packages/Caching/ValueObject/Storage/MemoryCacheStorage.php
0.78 seconds: [file] config/set/php80.php
0.69 seconds: [file] packages/PhpAttribute/AnnotationToAttributeMapper/ArrayItemNodeAnnotationToAttributeMapper.php
0.63 seconds: [file] packages/StaticTypeMapper/PhpDocParser/NullableTypeMapper.php
0.62 seconds: [file] rules/CodingStyle/Rector/ClassConst/VarConstantCommentRector.php
0.60 seconds: [file] packages/PhpAttribute/NodeAnalyzer/ExprParameterReflectionTypeCorrector.php
0.56 seconds: [file] packages/PhpAttribute/AnnotationToAttributeMapper/ArrayItemNodeAnnotationToAttributeMapper.php
0.45 seconds: [file] packages/Testing/PHPUnit/AbstractRectorTestCase.php
0.38 seconds: [file] rules/CodeQuality/Rector/ClassMethod/NarrowUnionTypeDocRector.php
0.35 seconds: [file] packages-tests/Comments/CommentRemover/CommentRemoverTest.php
0.34 seconds: [file] packages/PhpAttribute/Enum/DocTagNodeState.php
0.34 seconds: [file] packages/PostRector/Collector/NodesToAddCollector.php
0.32 seconds: [file] packages/StaticTypeMapper/PhpDocParser/IdentifierTypeMapper.php
0.31 seconds: [file] packages/BetterPhpDocParser/PhpDocNodeVisitor/TemplatePhpDocNode
...
```

Starting from here you can use your favorite profiler to analyse the slowest ones

-----

```php
<?php // parse.php
declare(strict_types=1);

// 
// inspired and adopted from https://gist.github.com/ruudk/41897eb59ff497b271fc9fa3c7d5fb27

$log = new SplFileObject("rector.log");

$logs = [];
$file = null;
while (! $log->eof()) {
    $line = trim($log->fgets());
    if ($line === '') {
        continue;
    }

    if (str_starts_with($line, '[file]')) {
        $file = $line;
        continue;
    }

    if ($file === null) {
        continue;
    }
    if (preg_match('/took (?<seconds>[\d.]+) s/', $line, $matches) === 1) {
        $accu = 0.0;
        if (array_key_exists($file, $logs)) {
            $accu = $logs[$file][0];
        }
        $logs[$file] = [$accu + ((float) $matches['seconds']), $file];
        $file = null;
    }
}

usort($logs, fn(array $left, array $right) => $right[0] <=> $left[0]);
$logs = array_slice($logs, 0, 100);

echo "Slowest files" . PHP_EOL;
foreach ($logs as $log) {
    echo sprintf("%.2f seconds: %s", $log[0], $log[1]) . PHP_EOL;
}
```

### CODE REVIEW

File: src/Rector/AbstractRector.php [hunk 1]
```diff
@@ -208,8 +209,13 @@ final public function enterNode(Node $node)
             $this->changedNodeScopeRefresher->reIndexNodeAttributes($node);
         }
 
+        $startTime = microtime(true);
+        $previousMemory = memory_get_peak_usage(true);
```
Comment (samsonasik) [inline src/Rector/AbstractRector.php]: this should only defined on debug mode, I prefer `$this->rectorOutputStyle->isDebug()` is saved to variable to ensure no repetitive calls, start with above `printDebugCurrentFileAndRule()` call...

File: src/Rector/AbstractRector.php [hunk 2]
```diff
@@ -220,6 +234,14 @@ final public function enterNode(Node $node)
             throw new ShouldNotHappenException($errorMessage);
         }
 
+        return $this->postRefactorProcess($originalNode, $node, $refactoredNode);
+    }
+
+    /**
+     * @param Node|Node[] $refactoredNode
+     */
+    private function postRefactorProcess(Node|null $originalNode, Node $node, Node|array $refactoredNode): Node
```
Comment (staabm) [inline src/Rector/AbstractRector.php:243]: extracted a method to reduce complexity of `enterNode()`

**Review by samsonasik — APPROVED** [review]

**Comment by TomasVotruba** [comment]
Thank you @staabm :partying_face:  I was thinking about this 2 days ago after your PRs, but you're faster :clap:

**Comment by staabm** [comment]
fwiw, I have already prepared a blog post, so people get an idea on how to use it:
https://github.com/staabm/staabm.github.io/pull/72

**Comment by TomasVotruba** [comment]
@staabm Amazing :clap:

**Comment by yguedidi** [comment]
@staabm great work! also though one day to do something like that 😅 
small remark, the logging being at the AbstractRector level, there will be log line per file **per rector**.
so maybe the parse.php snippet should be updated to either mention the slow rector, either sum the multiple logs for a file.
what do you think?

**Comment by staabm** [comment]
thanks for the feedback guys.

> small remark, the logging being at the AbstractRector level, there will be log line per file **per rector**.
> so maybe the parse.php snippet should be updated to either mention the slow rector, either sum the multiple logs for a file.
> what do you think?

I had the same idea while writing the script. I dropped it, as I think the goal of the script is to find the slow file.
Afterwards you use a real profiler with the very same rector config, so you reproduce the initial problem in a more realistic way.

its not important at that point which rector was slow. the profiler will tell you which part is most important. its the reason why you need the profiler. if we limit it to a certain rector we loose the overall picture.

**Pull request merged**